program-runtime/src/sysvar_cache.rs

```diff
@@ -26,189 +26,226 @@ impl ::solana_frozen_abi::abi_example::AbiExample for SysvarCache {
+
+    // deprecated sysvars, these should be removed once practical
     #[allow(deprecated)]
-    fees: Option<Arc<Fees>>,
-    rent: Option<Arc<Rent>>,
-    slot_hashes: Option<Arc<SlotHashes>>,
+    fees: Option<Fees>,
     #[allow(deprecated)]
-    recent_blockhashes: Option<Arc<RecentBlockhashes>>,
-    stake_history: Option<Arc<StakeHistory>>,
-    last_restart_slot: Option<Arc<LastRestartSlot>>,
+    recent_blockhashes: Option<RecentBlockhashes>,
 }
 
 impl SysvarCache {
-    pub fn get_clock(&self) -> Result<Arc<Clock>, InstructionError> {
-        self.clock
-            .clone()
-            .ok_or(InstructionError::UnsupportedSysvar)
-    }
-
-    pub fn set_clock(&mut self, clock: Clock) {
-        self.clock = Some(Arc::new(clock));
-    }
-
-    pub fn get_epoch_schedule(&self) -> Result<Arc<EpochSchedule>, InstructionError> {
-        self.epoch_schedule
-            .clone()
-            .ok_or(InstructionError::UnsupportedSysvar)
+    fn sysvar_id_to_buffer(&self, sysvar_id: &Pubkey) -> &Option<Vec<u8>> {
+        if *sysvar_id == Clock::id() {
+            &self.clock
+        } else if *sysvar_id == EpochSchedule::id() {
+            &self.epoch_schedule
+        } else if *sysvar_id == EpochRewards::id() {
+            &self.epoch_rewards
+        } else if *sysvar_id == Rent::id() {
+            &self.rent
+        } else if *sysvar_id == SlotHashes::id() {
+            &self.slot_hashes
+        } else if *sysvar_id == StakeHistory::id() {
+            &self.stake_history
+        } else if *sysvar_id == LastRestartSlot::id() {
+            &self.last_restart_slot
+        } else {
+            &None
+        }
     }
 
-    pub fn set_epoch_schedule(&mut self, epoch_schedule: EpochSchedule) {
-        self.epoch_schedule = Some(Arc::new(epoch_schedule));
+    pub fn read_sysvar_into(
+        &self,
+        sysvar_id: &Pubkey,
+        length: usize,
+        offset: usize,
+        out_buf: &mut [u8],
+    ) -> Result<(), InstructionError> {
+        if let Some(ref sysvar_buf) = self.sysvar_id_to_buffer(sysvar_id) {
+            if length == 0 {
+                return Err(InstructionError::InvalidInstructionData);
+            }
+
+            match length.checked_add(offset) {
+                Some(limit) if limit <= sysvar_buf.len() => (),
+                _ => return Err(InstructionError::InvalidInstructionData),
+            }
+
+            if length != out_buf.len() {
+                return Err(InstructionError::InvalidInstructionData);
+            }
+
+            if let Some(sysvar_slice) = offset
+                .checked_add(length)
+                .and_then(|limit| sysvar_buf.get(offset..limit))
+            {
+                out_buf.copy_from_slice(sysvar_slice);
+            } else {
+                return Err(InstructionError::InvalidInstructionData);
+            }
+
+            Ok(())
+        } else {
+            Err(InstructionError::UnsupportedSysvar)
+        }
```

Comment:
In this function and and `get_sysvar_obj`, when `sysvar_id_to_buffer` returns `None`, you return the error `UnsupportedSysvar`. Why not just hoist this error into `sysvar_id_to_buffer`?

Reply:
hmm, it just seemed cleaner to do it this way. since we want to take a reference to the `Vec<u8>` and only copy what we need from it, if `sysvar_id_to_buffer` returned a `Result<&Vec<u8>, _>`, we would have to unwrap the `Option` in `sysvar_id_to_buffer` anyway

Reply:
Ohh I see what you mean. Yeah, maybe what you have is fine then!